### ISSUE TITLE
Per-system hyperspace entry distance

### ISSUE THREAD

**Issue by accursedCursive** [endless-sky/endless-sky#3518]
Any system which finds itself the setting of large battles (e.g. the Korath battlefields) benefits greatly from having ships appear further from the centre of the system.
I emulated this by having massive opposing fleets appear from two planets, which I placed far from the system centre, hence me knowing that major battles work better with larger distances between the fleets.

The way I'd suggest implementing this is to have a "border" attribute specified next to the "habitable" and "belt" attributes.
If, in the future, the "sit still to travel" hyperspace system is replaced with "reach the system border to travel" (which is plausible, since it would make pirates less hopeless and replaces stop-start travel with zooming around), then having system borders already in place would be immensely useful.

Alternatively, to make hyperspace travel more customisable, the attribute could be called "hyperspace interference", and be applicable to any entity: Stars, planets, ships, even projectiles.
The attribute would allow entities to pull ships out of hyperspace within a specified range, and prevent them from entering hyperspace.
Or instead of a specified range, the number value associated with the attribute could be weighed against the power of the ship's hyperdrives, making the range flexible.

**Comment by Tadrix** [comment]
> "reach the system border to travel"
> zooming around

Jump in, back off a bit, jump out. You'd need "reach the system border and, facing your next destination, have the star behind you" system.

**Comment by Brick63** [comment]
this could be a serious change to the game and that is not my decision

i will however bring some points from the steam discussion this came from

1: piracy would be a viable strategy as you would be able to run or avoid much more powerful foes
2: massive battles would no longer be confined to mostly the center of the system

**Comment by Amazinite** [comment]
>If, in the future, the "sit still to travel" hyperspace system is replaced with "reach the system border to travel" (which is plausible, since it would make pirates less hopeless and replaces stop-start travel with zooming around)

https://github.com/endless-sky/endless-sky/issues/1094#issuecomment-220497460
>In EV each system only contained one or two planets. If ES were modified so that you can't jump while near a planet, getting away from all the planets would be complicated in some situations, and non-trivial for the AI to plan. (I actually tried adding in that constraint, a year or so ago, but concluded that it just made hyperspace jumping too annoying.)
>Both NAEV and OOlite make it considerably more difficult to jump between systems, and it can be quite annoying - especially when trying to cover a large distance quickly.

Same situation as having to reach the border of a system before jumping.

**Comment by Brick63** [comment]
based on all this i can say with a degree of certainty that this likely wont happen

**Comment by accursedCursive** [comment]
How is having to reach the system border before jumping even remotely difficult?
Ships already have code to achieve a specified position, orientation, and velocity, with as little or as much accuracy as desired and with the ability to have new goals every frame.
Unless that code is really hackish (I haven't checked it out yet) it should be easy to re-use it.

**Comment by jafdy** [comment]
It's not necessarily difficult, We just don't want to do it.

**Comment by accursedCursive** [comment]
Well, _someone_ should do it at some point.
I've emulated the effects of this feature by positionining planets that launch opposing fleets, and it massively improves systems that are scenes of large-scale combat.

**Comment by Amazinite** [comment]
>How is having to reach the system border before jumping even remotely difficult?

It was never stated to be difficult
>it just made hyperspace jumping too annoying

To change hyperspace mechanics this late in development is also a drastic change that is highly unlikely to happen.

>I've emulated the effects of this feature by positionining planets that launch opposing fleets, and it massively improves systems that are scenes of large-scale combat.

Not everything in this game is about large-scale combat, so no one is about to start making massive changes to the game mechanics that affect all aspects of the game while only subjectively improving one aspect.

**Comment by accursedCursive** [comment]
Hyperspace jumping isn't annoying enough.
You can travel anywhere, zero-risk, as long as you have fast steering.
You could make enemies with the Republic yet still travel from the Free Worlds to the deep of the Deep just by pressing J and waiting.
This ease of travel goes against ES lore, it dulls the gameplay, and takes the fun out of strategising your fleet movements since it doesn't matter in the slightest.

It would be much more interesting if the player had to make considerations like "Going through Acamar is the shortest route, but I have to cross the entire diameter of the system, and it's uncontrolled territory".
It could, in turn, make pirates seem a lot less ridiculous; instead of having them run into controlled systems and die, they could be shifted to uncontrolled systems where they can ambush anyone foolish enough to head through without a strong convoy.

Frankly, I don't see a single disadvantage to this change, being able to make large battles run more smoothly just seems like a bonus now.

**Comment by Amazinite** [comment]
>Hyperspace jumping isn't annoying enough.

The point of a game is to be entertaining, not annoying. 

>You can travel anywhere, zero-risk, as long as you have fast steering.
You could make enemies with the Republic yet still travel from the Free Worlds to the deep of the Deep just by pressing J and waiting.

There's a lot more to it that just "fast steering." Navy ships use weapons with lots of hit force, which hinders your ability to jump if hit, even if you do have fast turning. If you really want to have a ship that can travel from FW space to the end of the Deep with zero risk, then you're going to need to sacrifice a lot of outfit space to engines, a Scram Drive (and likely a Jump Drive as well), fuel pods, and ramscoops. That takes away from a ship's combat effectiveness, so it isn't like you can be highly combat effective AND travel anywhere with zero risk.

>This ease of travel goes against ES lore

Here's a tip: please stop talking about "ES lore," because you're just embarrassing yourself.

>Frankly, I don't see a single disadvantage to this change, being able to make large battles run more smoothly just seems like a bonus now.

The biggest disadvantage I see is that it will drastically slow down gameplay. Again, large battles are not everything. This would make EVERYTHING slower, from running jobs to trading goods, since this

>Acamar is the shortest route, **but I have to cross the entire diameter of the system**

makes it sounds like you want a NAEV system of hyperspacing (something that MZ specifically said he didn't like). It's already been stated that the hyperspace system will not be changed so that you need to stay away from stellar objects or go to the edge of the system, so it's not something you're going to win on. Honestly, if you deleted the last two paragraphs in the issue description, this would be entirely fine. (And to note, the [contributing page](https://github.com/endless-sky/endless-sky/blob/master/CONTRIBUTING.md) states that you should `Create a separate "issue" for each bug you are reporting and each feature you are requesting`, but you requested both wider system entry distances and tougher exit requirements, and the exit requirements you're requesting are what we are saying won't happen.)

**Comment by accursedCursive** [comment]
If you only want part of the idea, only take part of the idea.
Anyway, shouldn't it be the job of people who actually know C++ to choose which ideas to take?
In my experience, only forum warriors with no skills to speak of will get into a lengthly argument on the side of _not_ implementing a feature.

**Comment by Wrzlprnft** [comment]
It's a game design choice, not a feature. And the Person behind EndlessSky, Michael Zahniser, MZ, has said no, as quoted by Amazinite in his first comment. Not that you cannot convince him, but it seems you haven't done so, neither with one of the “people who actually know C++“ to give it a shot.

This is not a forum, and Amazinite is just using valid points you have not answered:
- a specialized ship will jump fast, but that is not the average player experience. have you considered that next to your own usual outfitting
- large battles are also not the average case, just jumping through is
- a straight up Border, where you either have just one point or line where you can jump, or a rather complex algorithm of “am i allowed to jump from here?“ because you are outside of the border, pointing in the right direction, but have the entire system in front of you, has been no-ed due to either annoyance or a significant junk being added to the Ai's “where shall i go“, which is computonally expensive.
- You posted two ideas, and one concrete solution, and the solution is what we are not happy with, and the “tougher exit“ idea has been debated an decided on, so unless you bring new ideas or points, that's off the table.

In short: by tying your propositions
- farther away system entry
- tougher system exit
into one issue, we cannot seperatly close them,
and by tieing it to one specific solution
- hyperspace system border
which has been no-ed, it is hard to constructively discuss parts of your idea without repeating ourselfs.

**Comment by accursedCursive** [comment]
Any competent programmer should know what I'm requesting is easy-peasy to implement (for someone familiar with ES code and the C++ language).
Except for a non-guessing method of finding the nearest point you can jump from if there are multiple objects that prevent you from jumping near them, that would actually take skill.

**Comment by Amazinite** [comment]
>Any competent programmer should know what I'm requesting is easy-peasy to implement

It's not about if this can by implemented or how difficult it would be to implement, it's about should we implement it and why should it be implemented. A deflection to "oh, you'll just know how to do it," doesn't give any reason as to why it should be implemented, it's only a deflection of the current criticism as to why it should not be implemented.

**Comment by accursedCursive** [comment]
Per-system hyperspace entry distance would be immensely useful to modders, and slightly useful to the vanilla game.
I had a problem just now with the starter-mission Bastion camping the planet I had to land on with an interdictor field, which wouldn't be a problem if I could encourage ships to move in a wider area.
I was barely able to make it, but I can easily see a newer player getting absolutely pwned in that situation.

By the way, hyperspace entry distances could be tied to the region in which ships are encouraged to stay within, since yesterday I found out that was a thing.

**Comment by KiLEdEnNis** [comment]
implementing this would make it very annoying for players to make hyperspace jumps, it may even be a reason for some players to quit the game, and either way it will be largely unappreciated by the playerbase
the "upside" is realism, stuff like bigger battles & whatever ya said is more of an opinion so we can't count that as upside either
another downside is the CPU would have a harder time, which decreases playability on low-end computers

the upsides don't outweigh the downsides even in the slightest which makes it not worth to implement

**Comment by accursedCursive** [comment]
Hyperspace jumps wouldn't be more annoying if just the core idea, adjustable hyperspace entry distances, were added.
There are no downsides to that, and that's the main thing I'm asking for.

**Comment by KiLEdEnNis** [comment]
that's fine, but your idea of creating a border ring where one can't jump if they're inside is 1 of the things that will not happen
being able to set an "on enter" distance per system would be a lot of work to do for each system, but one would enter closer to or further from the core in various systems which seems fine to me

**Comment by accursedCursive** [comment]
You set a default value, since most systems in unmodded ES don't need custom border sizes.
Even without a default value the work would be trivial, you just pick a number for every system that seems reasonable for how busy the system is.

Anyway, this really needs to be done, because 90% of the problems with playing my modded version of the game are related to hyperspace entry distances.
When you enter a system or leave a planet, there's a chance you'll be in short range of a powerful ship and die instantly.
Even without travelling or leaving planets, it's still a problem, I even had a Wasp warp on top of me and empty a Javelin Pod into me while I was flying a Sparrow.
Which works out to about 3000 damage to a ship with 1000 shields and 200 hull.

**Comment by Amazinite** [comment]
I remembered a discussion that occurred on the Discord a long time ago that related to the enter distance of systems, and this is the current behavior:
![screen shot 2018-01-12 at 5 45 57 am](https://user-images.githubusercontent.com/17688683/34871503-e40d8240-f75b-11e7-89f1-9515c04400a9.png)
Perhaps instead of adding a mechanic where each and every system can have its own entry distance, the simpler solution would just be to increase that 300 to 600 from destination enter distance to something much larger and wider, say 1000 to 2000, and have the angle that you land be close to the angle you entered (i.e. if a center system is connected to two other systems 180 degrees from one another, fleets entering from these two systems would end up roughly 180 degrees apart with respect to the target planet or system center, instead of the current situation where Angle::Random().Unit() seems to just throw your entry position anywhere 360 degrees around the target planet or system center).

**Comment by accursedCursive** [comment]
What problem is that "solution" even meant to solve?
It's bad for _everyone_.
It's bad for the vanilla game, because you're arbitrarily boosting the entry distance in a game with slow ships and short-range weapons.
It's bad for my version, because it seriously needs to be customisable, and a lazy solution like yours makes programmers decide they don't need a real solution.

**Comment by Amazinite** [comment]
>What problem is that "solution" even meant to solve?

How is this literally not what you've been asking for? You want fleets to spawn farther apart. That change would make fleets spawn farther apart. And simply because slow ships and short-range weapons exist, a change can't be made? How about the fast ships and long-range weapons that would now gain an upper hand against the mostly slow and powerful close-range heavy warships that a player must often fight, but is at a large disadvantage due to ships spawning so close to one another? And how arbitrary a change is is also an irrelevant point, as the current numbers are equally as arbitrary.

As for this being bad for the vanilla game, I'd suggest just the opposite. There have been many criticisms of how currently a player can be ganked just after departing or by pirates following them in the early game because pirates spawn VERY close to the player and VERY close to the planet. Both extending the enter distance away from the planet and stretching the range of that distance farther apart will give players more time to react once they have just departed and when entering a system, they will on average end up farther away from pirate fleets, which again gives the player more time to react. Such a change has its merits in the vanilla game. Yes, it isn't EXACTLY what you want, but that doesn't automatically mean that it is a bad change for everyone.

**Comment by KiLEdEnNis** [comment]
what @Amazinite proposed is a simple yet useful change that helps players react against instant-death situations, thus new players would die way less to RNG pirates, AND the result is in line with what you want

if you don't want this, try proposing something else; tell us why you don't want this (give actual reasons) and try to convince us with good arguments for your idea

what you're basically doing is "I want this feature because I want it", you ignore our arguments, and when we cooperate & try to help you just throw that away to try and get your way
this really will get you nowhere; this is not your game, it's MZ's game, there are a lot of people playing it and a lot of people working on it; learn to cooperate with them, learn to make a compromise that everyone is happy with, and think "do these upsides outweigh potential downsides?"

**Comment by Wrzlprnft** [comment]
For Protocoll, i was saying that it is processor expensive when you have to add a procedure for AI-ships to decide from where they are allowed to jump in relation to their current position. not that the implementation was hard as a task itself.

**Comment by kikotheexile** [comment]
You keep calling the dev and other contributors incompetent and keep flinging insults rather than giving us a reason to do the thing. 

Then, once someone figures out on their own why there might possibly be a good reason for your change (because you can't provide one) and comes to their own conclusion (based on your long-but-vague-and-insulting-rants that say nothing except "im the smartest person ever so do this thing i dont know how to do for me"). You then, in turn, say that your own idea is a bad one, goes against the lore, and breaks the game.

Like TF do you want? Seriously, tell us. Every time we try to help or work with, you just somehow manage to both ignore and go off on us at the same time. 

If you want us to do something then convince us and be willing to work with us. Calling professionals in the field that work on this free game for funsies incompetent isn't a good argument for your would-be feature.

**Comment by KiLEdEnNis** [comment]
so (based on my last comment)
- learn to cooperate with other contributors, don't talk down on anyone
- learn to make a compromise that everyone is happy with, give good arguments and listen to ours
- think "do these upsides outweigh potential downsides?", some ideas were no'd because gameplay > realism in those specific matters

if you change your attitude we're happy to have you amongst us

**Comment by accursedCursive** [comment]
The reason I'm complaining a lot is because I'm pretty sure most of you don't program at all, given how many of you are saying completely incorrect things, like Wrzecetera saying it's processor expensive for AI ships to find a jump point (It's actually just a few mathematical operations).

And I'm not going back on my idea, even though I rejected what Amazinite suggested.
-Current state: Entry distances hardcoded to 300-600
-What I want: Entry distances to not be hardcoded
-What Amazinite wants: Entry distances hardcoded to 1000-2000

**Comment by kikotheexile** [comment]
"Stop insulting people and be willing to work with them. Give us a good reason."
*immediately insults content creators and people who are working on this game, providing no argument and not giving a reason to do the thing*

I have one question and I challenge you to answer it: "To what end?"

**Comment by accursedCursive** [comment]
People who don't actually work with the game's source code should not be deciding whether an idea is feasible or not.
Which, I'm pretty sure, is no one in this thread, given that all the arguments started because people started making nonsense oppositions to the idea.

I tried to do a build of ES today to fix this issue myself, but the dependencies are an absolute mess, I couldn't even finish a compilation after following the instructions on the wiki and using a few hackish solutions to fix issues that the wiki didn't cover (such as making a directory with certain files in it to appease a GCC bug).
What I've got now is 50 compilation errors, one of them an undefined reference in AI.cpp and 49 of them being undefined references in Audio.cpp.
Maybe it doesn't work as cleanly on Windows 10, the wiki does mention that it's only been tried on Windows 7.

**Comment by Wrzlprnft** [comment]
i compiled it succesfully on win10.
yes, it is not processor intensive to find a point to jump from. Once. For one ship. But how often has that code to be executed? For how many ships?

**Comment by jafdy** [comment]
We decide if a feature is worth implementing. While we might have very different focuses we all want to make the best game possible, and that means working as a team, something that seems to be quite challenging for you. 
I will also say that many of us can work in with the code, but we are best elsewhere (eg. Wrzlprnft and Brick do art, and Amazinite does story). Just because you have one idea doesn't make it the best thing for the game. So far you have provided no reason to implement your idea and we have provided multiple gameplay reasons to not implement your idea, and you don't need to be able to code to make gameplay judgments.
I'm all for spreading ships out but I don't think this is the way to do it.

**Comment by kikotheexile** [comment]
Once again: "To what end?"

You never answered a three word question. You never gave us any good reason to implement your """"fix""""

All you've done is inflate your own ego and ignore literally everyone. How can you code if you cant even read? How can you be intelligent if you cant answer even the most simplistic and open ended question?

**Comment by endless-sky** [comment]
First of all, @accursedCursive, you should probably look at the [contributors](https://github.com/endless-sky/endless-sky/graphs/contributors) list before assuming that the people you're talking to don't know what they're talking about.

Secondly, if you're manually copying header files, etc. to other locations to fix what you think is a "GCC bug," you should consider the fact that maybe you just haven't set up your compilation environment correctly, and that all your "hackish" changes are just making the situation even worse.

**Comment by Disiuze** [comment]
I would like to say that the above comment made my day.

**Comment by Brick63** [comment]
I won't say I know how the code works I just want to know what all will be enhanced by this addition

**Comment by KiLEdEnNis** [comment]
@accursedCursive I'll say it yet again:
- change your attitude
- learn to work in a team
- don't talk down on others, especially since you haven't shown us anything to value your opinion
- know your place; MZ owns the game, and we're contributors helping each other in a nice way

if your attitude stays the same we have no place for you here
if you keep arguing the same way you've done; ignoring our arguments & giving no useful info back, your ideas won't have a place here

if you do change your attitude, argue properly, and treat the rest of us nice, we'll do the same back
if you have an idea which may be added but it needs some changes, we'll help and argue in a mature way to decide what the best solution is, together
and I think an apology to the rest of us seems in place too, since you basically trash-talked about us


so again: if you'll change for the better we'd accept you in the community, but if you stay like this, just leave

**Comment by accursedCursive** [comment]
It's impossible to work in a team if you hardcode loads of stuff and leave insufficient documentation.
For example, a good portion of outfit tags can only be discovered by content creators when they peruse the data files.
Plus, GCC is hardcoded to look for a particular file in a particular place (C:\Program Files\mingw64\x86_64-w64-mingw32\lib\libmingw32.a), instead of allowing the compiler to provide that file.
Unless the way I'm providing it is wrong, and it's instead just defaulting to that path, but that file is in a search directory provided by the linker and all the other files in the same directory load fine.

I know my programming environment is a total mess, I've been forced to put dev64 in C:/ as well because although setting one subfolder of it as a search directory allowed the bulk of the compilation to finish, at the end it would look for dev64 in C:/ and adjusting search directories for the compiler and linker did nothing.

Based on my several hours of experience with C++, I should set dev64/bin and dev64/lib as search directories for the linker since they're compiled files, and dev64/include as search directories for the compiler since they're uncompiled files.
But apparently that's wrong.

**Comment by kikotheexile** [comment]
You keep blaming your behavior on bad programming.

"I cant work with you because blah blah blah is hard coded" is a bullshit answer and you know it. You and your "several hours" of experience need to find your chill and be willing to read our responses. Tell us exactly what you want, and why it's a good idea.

**Comment by kikotheexile** [comment]
There are a lot of systems. Giving each one a unique jump distance would be a pain. That said, i could see adding system attributes and leaving a blank attribute field to default.

How would this improve gameplay?

**Comment by KiLEdEnNis** [comment]
> I've already said what I want, and why it's a good idea.

basically why it's a good idea: "it's my idea so it's good"
we have good arguments against, the response: "no, it's my idea, I say so, and I'm right", then you just ignore the rest, and I'm not willing to repeat half the discussion, just read back

we keep trying to work with you but you keep making up dumb excuses, and honestly, we're tired of your attitude
frankly, you and your several hours of C++ experience should just get out, since all you can do is trash talk, act like a know-it-all and defend your own point instead of looking for a real answer

**Comment by warp-core** [comment]
@accursedCursive If you want help with figuring out how to compile ES, come to the discord and I'll see if I can figure out what you're doing wrong and how to fix it.

**Comment by MCOfficer** [comment]
unless i'm a sloppy reader, i didn't see you picking up many of the counter-arguments, though. it seems to me that you are mainly repeating your pro-system-border argument, no matter what someone replies to you.
not to mention that you immediately dismissed the only attempt to work towards your idea someone (@Amazinite ) made.

to be quite frank:
the way you appear to other people is stubborn, narcisstic and intent on your own and only your own ideas. that's why people are annoyed and even try to give you tips how to behave more cooperatively.

maybe now you understand why you are getting the replies you get.

**Comment by Brick63** [comment]
bulk freighters are meant to sacrifice speed for cargo so 20 seconds would be a trade off
also quarg only have heavy warships try doing some research before speaking

**Comment by accursedCursive** [comment]
In my version of Endless Sky, the Quarg have more ships and outfits.
Like Roadrunner, a high-speed interceptor, and whatever I'm going to call this fighter I'm making that's really just a smaller Roadrunner like the Finch is just a smaller Sparrow.

**Comment by Brick63** [comment]
well this is base game so non base game things should not be used to try and legitimize/ delegitimize stuff for the base game

**Comment by KiLEdEnNis** [comment]
> In my version of Endless Sky, the Quarg have more ships and outfits.

that's not the base game, don't expect us to know it
so as Brick said: do some research before speaking, in this case research what's mod content & what's base

anyway, initially you straight-up no'd Amazinite's idea:
> Perhaps instead of adding a mechanic where each and every system can have its own entry distance, the simpler solution would just be to increase that 300 to 600 from destination enter distance to something much larger and wider, say 1000 to 2000,

even though it seems like a good change to prevent new players from dying to rng
as for your arguments agains: a Bulk Freighter is 1 of, if not the slowest ship in the game, not doing a balance change because very heavy slow freighters would take a little while is illogical; if you want speed get smaller freighters

having a unique entry distance per system is a big change and might look annoying to players; where in some systems they enter close to the star, and sometimes far away from it, I don't think that this change is worth it

**Comment by kikotheexile** [comment]
If ships you are using in your mod are too slow for you, change their speed in your mod. Its your mod.

That said, i dont have any problem with adding system attributes. And attributes to diffwrent types of drives to make different ships behave differently. I just also dont see a need to completely unlock jump in rings, as this causes more problems than solves

**Comment by accursedCursive** [comment]
>having a unique entry distance per system is a big change and might look annoying to players; where in some systems they enter close to the star, and sometimes far away from it, I don't think that this change is worth it
Unless it's taken to extremes, players aren't likely to notice the difference, even if it does improve their gameplay.
It's very common for game developers to make subtle changes to a game that make it significantly better.

>I just also dont see a need to completely unlock jump in rings, as this causes more problems than solves
What problems?

**Comment by MCOfficer** [comment]
> What problems?

did you even _read_ our comments.

**Comment by tehhowch** [comment]
I'll just mention a few things:
 - the 300-600 entry ring is only for jump drive entry into a system. Arguably, it could be increased to 400-800 to match the similar bits in the Fleet class.
 - the entry rings are not uniform distributions (they are biased towards lower radius, which makes some sense as that is where a significant object of interest is. Uniform sampling of a circular area requires taking the square root of the `Random::Real` result. Switching to a uniform sampling would result in an small increase in the average fleet distance when entering)
 - I do not see "inconvenienced hyperspacing" being added anytime soon, per linked comments above from MZ
 - allowing each system to specify a planetary offset (instead of the default "action radius" of 400 currently used) and referencing this offset when spawning ships is not difficult.
   - I don't see this getting used often, with the sole exception of a mission or two which involves searching a number of planets in a single system, and that can already be effectively done (more, really) by piping the player into that system via wormhole only, since they won't exit hyperspace and thus won't be facing a landable planet.

Anyone looking to implement this will need to choose a variable name and give it a default value in System.h, choose a node keyword to indicate that is the value to be modified in System:: Load, add a public method to access the value, and then modify the relevant places in Engine and Fleet classes to pull in this value instead.

**Comment by accursedCursive** [comment]
I should add that I'm aware that entry distances are currently measured from the planet, even though I'm suggesting an entry distance measured from the system centre.
Which would mean the new system would either abandon directing ships at the planet (which would require changing a lot of code), or the destinations would be intially calculated the same way but if the destination is in the no-hyperspace-entry zone it's pulled back.
The latter sounds much easier to implement, since only the ship class would need to be changed.

As for Jump Drives, any suggestions?

**Comment by endless-sky** [comment]
@accursedCursive, you can just modify the CodeBlocks project settings to point to wherever you installed MinGW and the dev64 libraries. No need at all to go copying files around trying to "fix" things. Probably all you've achieved by doing that is to muck up your system.

**Comment by accursedCursive** [comment]
Never knew project settings were a thing, my compiling is fine now.
Except that Audio.cpp still throws dozens of "undefined reference to _imp__somethingsomething" errors.

**Comment by Amazinite** [comment]
Oh, hey, this issue was a thing.
Arrival distances are a thing now. #4432
